sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusSessionManager.java

```diff
@@ -295,6 +301,9 @@ Mono<ServiceBusReceiveLink> getActiveLink() {
                 } else if (failure instanceof AmqpException
                     && ((AmqpException) failure).getErrorCondition() == AmqpErrorCondition.TIMEOUT_ERROR) {
                     return Mono.delay(SLEEP_DURATION_ON_ACCEPT_SESSION_EXCEPTION);
+                } else if (failure instanceof AmqpException
+                    && ((AmqpException) failure).isTransient()) {
```

Comment:
If we agree to keep the _retryWhen_ as it is, then we could adjust the above code comment to something like -

_While waiting for the link to ACTIVE, if the broker detaches the link without an error condition, the link-endpoint-state publisher will transition to completion without ever emitting ACTIVE. Map such publisher completion to transient (i.e., retriable) AmqpException to enable processor recovery._

Reply:
Thanks @anuchandy, I also thought a lot for whether to add this transient error check to the link-level retry. 

I was refering [the error handling of ServiceBusReceiveLinkProcessor (SBRLP)](https://github.com/Azure/azure-sdk-for-java/blob/62d29b71ae2a8485dae1f3fa14e40f37fdcc2d4a/sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReceiveLinkProcessor.java#L340-L357), when it is a transient error, SBRLP will retry to request a new link rather than throw to higher level. On the other hand, if we throw the transient error to SessionProcessor, the processor will restart the client as a recovery, not sure if it is too heavy. So I decide to also add a transient error check here for session link. 

However, for this specific case, the error transient flag for the error is set by us. Actually adding this transient error check may change the existing behavior and bring in unexpected error. Maybe I should create another issue to add the transient check and do the validation. For this PR, I can let the session processor do the recovering as it has less impact.

Reply:
I don't have context about the current design choice of propagating the error to the session processor, may be inherited from the legacy library.

Reply:
But I suspect the reason for the design could be - the SB session concept allows load balancing of the client-side processing by distributing the messages across sessions. It could be that the typical case is to spread the messages across sessions evenly to maximize such parallelism. So, suppose one session disconnect due to no message, no message read (i.e., idle-timeout),  service busy, service update, or connection error. In that case, it's likely that other sessions to the same entity are going be disconnected for the same reason. This common use case might be one reason for the current design choice, which simplifies the recovery by avoiding the need to coordinate all those threads across sessions trying to recover locally.

Reply:
Sure, I'll update the code comment. 

I found that if we throw the error to session processor level, because the recover process need to close client, which contains the block calls, and the link-endpoint-state publisher emit the signal on reactor-exector thread, the session processor recovery would fail with an error "block()/... is not allowed on the reactor-exector thread".

This `getActiveLink()` is also used for `ServiceBusSessionReceiverAsyncClient` to accept a session, and user could handle if an error throw out. But for the session processor, I suspect when we reused the `getActiveLink()`, we didn't consider this error path and validate it. We have to publish the `Mono.<Long>error(failure)` on a blocking-allowed thread so that it can be handled by the session processor.